File: prayer_app_expo/src/models/prayerTimes.js
// Cleans timezone suffix from API time strings like "05:43 (EET)"
export function cleanTime(raw) {
    if (!raw) return '00:00';
    return raw.replace(/\s*\(.*\)/, '').trim();
}

// Converts "17:49" → "5:49 PM"
export function formatTo12Hour(time24) {
    const parts = time24.split(':');
    if (parts.length < 2) return time24;
    let h = parseInt(parts[0], 10) || 0;
    const m = parts[1];
    const ampm = h >= 12 ? 'PM' : 'AM';
    if (h === 0) h = 12;
    else if (h > 12) h -= 12;
    return `${h}:${m} ${ampm}`;
}

// Parses "HH:MM" into a Date on the given date
export function timeToDate(time24, date) {
    const parts = time24.split(':');
    const h = parseInt(parts[0], 10) || 0;
    const m = parseInt(parts[1], 10) || 0;
    const d = new Date(date);
    d.setHours(h, m, 0, 0);
    return d;
}

// Parse API response into a PrayerTimings object
export function parsePrayerTimings(json) {
    const data = json.data;
    const timings = data.timings;
    const hijri = data.date?.hijri;
    const greg = data.date?.gregorian;

    const gregorianFormatted = greg
        ? `${greg.weekday?.en || ''}, ${greg.month?.en || ''} ${greg.day || ''}, ${greg.year || ''}`
        : '';

    return {
        fajr: cleanTime(timings.Fajr),
        sunrise: cleanTime(timings.Sunrise),
        dhuhr: cleanTime(timings.Dhuhr),
        asr: cleanTime(timings.Asr),
        maghrib: cleanTime(timings.Maghrib),
        isha: cleanTime(timings.Isha),
        lastThird: cleanTime(timings.Lastthird),
        hijriDay: hijri?.day || '',
        hijriMonthAr: hijri?.month?.ar || '',
        hijriYear: hijri?.year || '',
        gregorianFormatted,
        get hijriFormatted() {
            if (!this.hijriDay) return '—';
            return `${this.hijriDay} ${this.hijriMonthAr} ${this.hijriYear} هـ`;
        },
        get mainPrayers() {
            return [
                { name: 'Fajr', time24: this.fajr, time12: formatTo12Hour(this.fajr), isMain: true },
                { name: 'Dhuhr', time24: this.dhuhr, time12: formatTo12Hour(this.dhuhr), isMain: true },
                { name: 'Asr', time24: this.asr, time12: formatTo12Hour(this.asr), isMain: true },
                { name: 'Maghrib', time24: this.maghrib, time12: formatTo12Hour(this.maghrib), isMain: true },
                { name: 'Isha', time24: this.isha, time12: formatTo12Hour(this.isha), isMain: true },
            ];
        },
        get supplementaryPrayers() {
            return [
                { name: 'Sunrise', time24: this.sunrise, time12: formatTo12Hour(this.sunrise), isMain: false },
                { name: 'Last Third of Night', time24: this.lastThird, time12: formatTo12Hour(this.lastThird), isMain: false },
            ];
        },
    };
}

// Returns the next upcoming main prayer
export function getNextPrayer(timings, now) {
    const prayers = timings.mainPrayers;
    for (const p of prayers) {
        const t = timeToDate(p.time24, now);
        if (t > now) return p;
    }
    return prayers[0]; // Tomorrow's Fajr
}

// Duration (ms) until the next prayer
export function getTimeUntilNext(timings, now) {
    const next = getNextPrayer(timings, now);
    if (!next) return 0;
    let target = timeToDate(next.time24, now);
    if (target <= now) {
        target = new Date(target.getTime() + 24 * 60 * 60 * 1000);
    }
    return target.getTime() - now.getTime();
}

// Format milliseconds as "HH:MM:SS"
export function formatCountdown(ms) {
    if (ms <= 0) return '00:00:00';
    const totalSec = Math.floor(ms / 1000);
    const h = String(Math.floor(totalSec / 3600)).padStart(2, '0');
    const m = String(Math.floor((totalSec % 3600) / 60)).padStart(2, '0');
    const s = String(totalSec % 60).padStart(2, '0');
    return `${h}:${m}:${s}`;
}

File: prayer_app_flutter/lib/src/models/prayer_times.dart
/// Cleans timezone suffix from API time strings like "05:43 (EET)"
String cleanTime(String? raw) {
  if (raw == null || raw.isEmpty) return '00:00';
  return raw.replaceAll(RegExp(r'\s*\(.*\)'), '').trim();
}

/// Converts "17:49" → "5:49 PM"
String formatTo12Hour(String time24) {
  final parts = time24.split(':');
  if (parts.length < 2) return time24;
  int h = int.tryParse(parts[0]) ?? 0;
  final m = parts[1];
  final ampm = h >= 12 ? 'PM' : 'AM';
  if (h == 0) {
    h = 12;
  } else if (h > 12) {
    h -= 12;
  }
  return '$h:$m $ampm';
}

/// Parses "HH:MM" into a DateTime on the given date
DateTime timeToDateTime(String time24, DateTime date) {
  final parts = time24.split(':');
  final h = int.tryParse(parts[0]) ?? 0;
  final m = int.tryParse(parts.length > 1 ? parts[1] : '0') ?? 0;
  return DateTime(date.year, date.month, date.day, h, m);
}

class PrayerEntry {
  final String name;
  final String time24;
  final bool isMain; // true for the 5 fard prayers

  const PrayerEntry({
    required this.name,
    required this.time24,
    this.isMain = true,
  });

  String get time12 => formatTo12Hour(time24);
}

class PrayerTimings {
  final String fajr;
  final String sunrise;
  final String dhuhr;
  final String asr;
  final String maghrib;
  final String isha;
  final String lastThird;
  final String hijriDay;
  final String hijriMonthAr;
  final String hijriYear;
  final String gregorianFormatted;

  const PrayerTimings({
    required this.fajr,
    required this.sunrise,
    required this.dhuhr,
    required this.asr,
    required this.maghrib,
    required this.isha,
    required this.lastThird,
    required this.hijriDay,
    required this.hijriMonthAr,
    required this.hijriYear,
    required this.gregorianFormatted,
  });

  factory PrayerTimings.fromApiResponse(Map<String, dynamic> json) {
    final data = json['data'];
    final timings = data['timings'] as Map<String, dynamic>;
    final hijri = data['date']?['hijri'] as Map<String, dynamic>?;
    final greg = data['date']?['gregorian'] as Map<String, dynamic>?;

    String gregFormatted = '';
    if (greg != null) {
      gregFormatted =
          '${greg['weekday']?['en'] ?? ''}, ${greg['month']?['en'] ?? ''} ${greg['day'] ?? ''}, ${greg['year'] ?? ''}';
    }

    return PrayerTimings(
      fajr: cleanTime(timings['Fajr'] as String?),
      sunrise: cleanTime(timings['Sunrise'] as String?),
      dhuhr: cleanTime(timings['Dhuhr'] as String?),
      asr: cleanTime(timings['Asr'] as String?),
      maghrib: cleanTime(timings['Maghrib'] as String?),
      isha: cleanTime(timings['Isha'] as String?),
      lastThird: cleanTime(timings['Lastthird'] as String?),
      hijriDay: hijri?['day'] ?? '',
      hijriMonthAr: hijri?['month']?['ar'] ?? '',
      hijriYear: hijri?['year'] ?? '',
      gregorianFormatted: gregFormatted,
    );
  }

  /// Parse a single day entry from the calendar API response.
  /// Each entry in the calendar `data[]` array has the same structure
  /// as the single-day response's `data` field.
  factory PrayerTimings.fromCalendarDay(Map<String, dynamic> dayData) {
    return PrayerTimings.fromApiResponse({'data': dayData});
  }

  String get hijriFormatted {
    if (hijriDay.isEmpty) return '—';
    return '\u200E$hijriDay $hijriMonthAr $hijriYear هـ';
  }

  /// Main 5 prayers in schedule order (matching Figma)
  List<PrayerEntry> get mainPrayers => [
        PrayerEntry(name: 'Fajr', time24: fajr),
        PrayerEntry(name: 'Dhuhr', time24: dhuhr),
        PrayerEntry(name: 'Asr', time24: asr),
        PrayerEntry(name: 'Maghrib', time24: maghrib),
        PrayerEntry(name: 'Isha', time24: isha),
      ];

  /// Supplementary times shown below divider
  List<PrayerEntry> get supplementaryPrayers => [
        PrayerEntry(name: 'Sunrise', time24: sunrise, isMain: false),
        PrayerEntry(name: 'Last Third of Night', time24: lastThird, isMain: false),
      ];

  /// Returns the next upcoming main prayer, or null if all passed
  PrayerEntry? getNextPrayer(DateTime now) {
    for (final entry in mainPrayers) {
      final t = timeToDateTime(entry.time24, now);
      if (t.isAfter(now)) return entry;
    }
    return mainPrayers.first; // Tomorrow's Fajr
  }

  /// Countdown duration until the next prayer
  Duration getTimeUntilNext(DateTime now) {
    final next = getNextPrayer(now);
    if (next == null) return Duration.zero;
    var target = timeToDateTime(next.time24, now);
    if (!target.isAfter(now)) {
      target = target.add(const Duration(days: 1)); // Tomorrow
    }
    return target.difference(now);
  }

  /// Apply per-prayer minute offsets. Returns a NEW PrayerTimings with adjusted times.
  /// Only main 5 prayers are offset; Sunrise and LastThird stay unchanged.
  PrayerTimings applyOffsets(Map<String, int> offsets) {
    String adjustTime(String time24, int minutes) {
      if (minutes == 0) return time24;
      final now = DateTime.now();
      final dt = timeToDateTime(time24, now).add(Duration(minutes: minutes));
      return '${dt.hour.toString().padLeft(2, '0')}:${dt.minute.toString().padLeft(2, '0')}';
    }

    return PrayerTimings(
      fajr: adjustTime(fajr, offsets['Fajr'] ?? 0),
      sunrise: sunrise,
      dhuhr: adjustTime(dhuhr, offsets['Dhuhr'] ?? 0),
      asr: adjustTime(asr, offsets['Asr'] ?? 0),
      maghrib: adjustTime(maghrib, offsets['Maghrib'] ?? 0),
      isha: adjustTime(isha, offsets['Isha'] ?? 0),
      lastThird: lastThird,
      hijriDay: hijriDay,
      hijriMonthAr: hijriMonthAr,
      hijriYear: hijriYear,
      gregorianFormatted: gregorianFormatted,
    );
  }

  /// Check that adjusted times are in valid order:
  /// Fajr < Sunrise < Dhuhr < Asr < Maghrib < Isha
  bool sanityCheck() {
    final order = [fajr, sunrise, dhuhr, asr, maghrib, isha];
    int prev = -1;
    for (final t in order) {
      final parts = t.split(':');
      if (parts.length < 2) return false;
      final mins = (int.tryParse(parts[0]) ?? 0) * 60 + (int.tryParse(parts[1]) ?? 0);
      if (mins <= prev) return false;
      prev = mins;
    }
    return true;
  }
}
